### PR TITLE
fix: template warning

### PR DESCRIPTION
**Pull request by c2tarun**
*Issue #, if available:*

*Why is this change necessary?*

*How does it address the issue?*

*What side effects does this change have?*

*Did you change a dependency in `requirements/base.txt`?*
    *If so, did you run `make update-reproducible-reqs`*

*Checklist:*

- [ ] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [x] Write/update integration tests
- [x] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/commands/_utils/options.py [hunk 1]
```diff
@@ -53,6 +53,13 @@ def get_or_default_template_file_name(ctx, param, provided_value, include_build)
         # sam configuration file should always be relative to the supplied original template and should not to be set
         # to be .aws-sam/build/
         setattr(ctx, "samconfig_dir", os.path.dirname(original_template_path))
+        try:
+            template_data = get_template_data(result)
+            setattr(ctx, "template_dict", template_data)
```
Comment (jfuss) [inline samcli/commands/_utils/options.py:58]: We should not be doing setattr directly. Can we not follow the same pattern as profile or region which have explicit setter methods?

Comment (sriram-mv) [reply samcli/commands/_utils/options.py:58]: https://github.com/aws/aws-sam-cli/blob/develop/samcli/cli/options.py#L33, something like that makes sense.

File: samcli/lib/warnings/sam_cli_warning.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    """
+    Top level class which all warnings should extend from.
+    """
+
+    def check(self, template_dict):
+        raise Exception("NotImplementedException")
+
+
+def _get_deployment_preferences_status(function):
+    deployment_preference = function["Properties"].get("DeploymentPreference", None)
```
Comment (sriram-mv) [inline samcli/lib/warnings/sam_cli_warning.py:19]: can we add a docstring for this?

File: samcli/lib/warnings/sam_cli_warning.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    def check(self, template_dict):
+        raise Exception("NotImplementedException")
+
+
+def _get_deployment_preferences_status(function):
+    """
+    Takes a AWS::Serverless::Function resource and checks if resource have a deployment preferences applied
+    to it. If DeploymentPreference found then it returns its status if it is enabled or not.
+    """
+    deployment_preference = function["Properties"].get("DeploymentPreference", None)
```
Comment (sriram-mv) [inline samcli/lib/warnings/sam_cli_warning.py:23]: should we do function.get("Properties", {}).get("DeploymentPreference", None) just to avoid failures.

**Review by sriram-mv — COMMENTED** [review]
This looks okay to me, we should come back and aim to fix the line you mentioned. we are now making two calls for telemetry instead of one, I dont think it will cause issues becuase timeouts are built-in. Can we check that there is no performance impact?

File: samcli/commands/_utils/options.py [hunk 4]
```diff
@@ -53,6 +53,14 @@ def get_or_default_template_file_name(ctx, param, provided_value, include_build)
         # sam configuration file should always be relative to the supplied original template and should not to be set
         # to be .aws-sam/build/
         setattr(ctx, "samconfig_dir", os.path.dirname(original_template_path))
+        try:
+            # FIX-ME: figure out a way to insert this directly to sam-cli context and not use click context.
```
Comment (awood45) [inline samcli/commands/_utils/options.py:57]: Can we close out the FIXME here?

**Review by awood45 — CHANGES_REQUESTED** [review]
This generally looks good - can we either resolve or remove the `FIX-ME` item?

**Review by awood45 — APPROVED** [review]

**Pull request merged**